CONTRIBUTING.md

```diff
@@ -1,103 +1,118 @@
 ```
 
-Install from pip:
+To load `pyenv` automatically, add the following lines to your shell's startup script, 
+such as `~/.bashrc` or `~/.zshrc` (and then either restart the shell or `source` the changed script):
 
 ```shell
-pip install --upgrade git+git://github.com/wandb/client.git
+export PYENV_ROOT="$HOME/.pyenv"
+export PATH="$HOME/.pyenv/bin:$PATH"
+eval "$(pyenv init --path)"
+eval "$(pyenv virtualenv-init -)"
 ```
 
-Or from pypi:
+Then run the following command to set up your environment:  
 
 ```shell
-pip install --upgrade wandb
+./tools/setup_dev_environment.py
 ```
 
-## Code organization
-
-```bash
-wandb/
-├── ...
-├── apis/   # Public api (still has internal api but this should be moved to wandb/internal)
-│   ├── ...
-│   ├── internal.py
-│   ├── ...
-│   └── public.py
-├── cli/    # Handlers for command line functionality
-├── ...
-├── integration/    # Third party integration
-│   ├── fastai/
-│   ├── gym/
-│   ├── keras/
-│   ├── lightgbm/
-│   ├── metaflow/
-│   ├── prodigy/
-│   ├── sacred/
-│   ├── sagemaker/
-│   ├── sb3/
-│   ├── tensorboard/
-│   ├── tensorflow/
-│   ├── torch/
-│   ├── xgboost/
-│   └── ...
-├── ...
-├── proto/  # Protocol buffers for inter-process communication and persist file store
-├── ...
-├── sdk/    # User accessed functions [wandb.init()] and objects [WandbRun, WandbConfig, WandbSummary, WandbSettings]
-│   ├── backend/    # Support to launch internal process
-│   ├── ...
-│   ├── interface/  # Interface to backend execution
-│   ├── internal/   # Backend threads/processes
-│   └── ...
-├── ...
-├── sweeps/ # Hyperparameter sweep engine (see repo: https://github.com/wandb/sweeps)
-└── ...
-```
+In particular, this tool will set up [`tox`](https://github.com/tox-dev/tox), which we use 
+for automating development tasks such as code linting and testing. 
 
-## Setup development environment
+## Building protocol buffers
 
-In order to run unittests please install pyenv:
+We use [protocol buffers](https://developers.google.com/protocol-buffers) to communicate 
+from the user process to the `wandb` backend process.
 
-(put the output in your `~/.bashrc`)
+If you update any of the `.proto` files in `wandb/proto`, you'll need to run:
 
 ```shell
-curl https://pyenv.run | bash
+make proto
 ```
 
-then run:
+## Linting the code
+
+We use [`black`](https://black.readthedocs.io/), [`flake8`](https://flake8.pycqa.org/), 
+and [`mypy`](http://mypy-lang.org/) for code formatting and checks (including static type checks). 
 
+To reformat the code, run: 
 ```shell
-./tools/setup_dev_environment.sh
+tox -e format
 ```
 
-## Building protocol buffers
-
-We use protocol buffers to communicate from the user process to the wandb backend process.
-
-If you update any of the .proto files in `wandb/proto`, you'll need to run:
+To run checks, execute: 
 
 ```shell
-make proto
+tox -e flake8,mypy
 ```
 
-## Code checks
-
-- Reformat: `tox -e format`
-- Type check: `tox -e flake8,mypy`
-- Misc: `tox`
-
 ## Testing
```

Comment:
what do you think of moving the `testing` section into a different readme under the `tests` folder and just link to it in here with a shorter description?

Reply:
I think this whole document should be converted into a dedicated section on our web site and CONTRIBUTING.md would simply point at it. See for example, how `numpy` people do that: https://numpy.org/devdocs/dev/index.html. 
However, I think it would make most sense in the future and for now I would just keep everything in one place.

Reply:
Ah I see! I like this Idea! So maybe TOC would be helpful, since the doc is pretty long. What do you think?

Reply:
Do you think this is not enough? I agree that it's not particularly verbose though...
![Kapture 2021-11-30 at 16 38 01](https://user-images.githubusercontent.com/7557205/144150733-0239aa17-9e89-460a-9703-764b59e867f9.gif)

Reply:
Ah that's actually great. I just wasn't aware of it :)

Reply:
Let's actually add it, that would be more verbose.

Reply:
https://ecotrust-canada.github.io/markdown-toc/ is great BTW

Reply:
ah that's cool!!!